### ISSUE TITLE
Check upgradable packages on OBS

### ISSUE THREAD

**Issue by pandom79** [openSUSE/open-build-service#11646]
**Is your feature request related to a problem? Please describe.**
```
When I go in my home project on OBS, i don't know if I have to upgrade my packages and thus
I have to check manually if a new version exists.
```
**Describe the solution you'd like**
```
For all packages for which a service is defined, 
we could check if a new version exists and labeled it as "upgradable" just like for "link"
```
**Describe alternatives you've considered**
```
From service file, we could point to URL and check if a new version exists via "osc check-upgrade". (just an example)
If we pass a prj as argument then it will check all packages 
   osc check-upgrade prj
If we pass a prj and a pkg as argument then it only will check the pkg
   osc check-upgrade prj pkg
```
**Additional context**
```
If this feature already exists, please ignore it
```

**Comment by pandom79** [comment]
> I doubt that we will ever do this in the main service because this is about the package content, we provide various ways to interact with the service for such things. Like osc has a [plugin concept](https://en.opensuse.org/openSUSE:OSC_plugins) and there are services like https://repology.org

I mean that when i go in my home project https://build.opensuse.org/project/show/home:pandom79 would be nice see next to name of a package the label "upgradable" as so as we see "link". I don't know obs and its api , for this reason i'm asking you if it could be done.

It has not to make an upgrade but only check that a new version exists and labeled as upgradable.
The user will have to upgrade, obviously. It would be such as a notification service.

**Comment by pandom79** [comment]
I written an osc service in Python that finds the newer versions of a package creating a file called "check_upgrade" which contains all newer versions. In this way, frontend side, when loading our home project could check if that file exists for each package and labeled it as "upgradable".

The following there are any examples :

```
<service name="check_upgrade" mode="disabled">
    <param name="urlsrc">https://archive.eclipse.org/eclipse/downloads/</param>
    <param name="regexurl">^drops[\w]*/R-[\d]+[.][\w]*[.]*[\w]*-[\w]+/$</param>
    <param name="regexver">[\d]+[.][\w]*[.]*[\w]*</param>
    <param name="currentver">4.7.3</param>
</service>

<service name="check_upgrade" mode="disabled">
    <param name="urlsrc">https://github.com/openSUSE/open-build-service/releases</param>
    <param name="regexurl">^/openSUSE/open-build-service/archive/refs/tags/[\d].[\d]+.[\d]+.tar.gz$</param>
    <param name="regexver">[\d].[\d]+.[\d]+</param>
    <param name="currentver">2.8.0</param>
</service>

<service name="check_upgrade" mode="disabled">
    <param name="urlsrc">https://serviio.org/download</param>
    <param name="regexurl">^https://download.serviio.org/releases/serviio-[\d]+.[\w]*[.]*[\w]*-linux.tar.gz$</param>
    <param name="regexver">[\d]+.[\w]*[.]*[\w]*</param>
    <param name="currentver">1.3</param>
</services>

<service name="check_upgrade" mode="disabled">
    <param name="urlsrc">https://pypi.org/project/github-api-python/#history</param>
    <param name="regexurl">^/project/github-api-python/[\d]*.[\w]*.[\w]*/$</param>
    <param name="regexver">[\d]+.[\w]*.[\w]*</param>
    <param name="currentver">0.1.2a</param>
 </service>

 <service name="check_upgrade" mode="disabled">
    <param name="urlsrc">https://www.jetbrains.com/pycharm/whatsnew/</param>
    <param name="regexurl">/pycharm/whatsnew/[\d]+-[\d]+/</param>
    <param name="regexver">[\d]+-[\d]+</param>
    <param name="currentver">2019-1</param>
    <param name="separator">-</param>
 </service>

```

The service uses a Html Parser and according the regular expression extract the versions which will be compared with the current version to determine the newer

**Github and Gitlab**:
For both Python offers the specific api but that means that it will be limited to them. For this reason I preferred to use a Html parser

**Comment by adrianschroeter** [comment]
hm, it would make more sense as a project wide service IMHO. 

It would need to find the remote URLs and parameters in the packages. Can't you parse things like the remoteurl from the eg. the spec files?

I wonder how this service could be executed, maybe as local only for now. That way you can interact with the packager when he is touch his packages.

Does this make sense?

**Comment by pandom79** [comment]
> hm, it would make more sense as a project wide service IMHO.
> 
I agree, I'd see fine it such as a background job, runs it for all packages for which is defined 
> It would need to find the remote URLs and parameters in the packages. Can't you parse things like the remoteurl from the eg. the spec files?
> 
Of course, I have the outdir parameter value which contains the package folder path....thus, the spec file too.
I can access to it and read.
The remote url should contain the web page address which contains the data that we need but not always is so.
See the jetbrains case which required "what's new" page, not download page.
The parser finds whatever you want not only links (href) 

> I wonder how this service could be executed, maybe as local only for now. That way you can interact with the packager when he is touch his packages.
> 
> Does this make sense?

Yep, at the moment it works so:
The packager create the package and related services including "check_upgrade".
Front-end side, when we load the home project, for each package, we have to check if a file called "check_upgrade" exists.
If exists, label it as "upgradable" just like we see "link" for example.
In this way, a user is notified and then he can go to inside package and see the check_upgrade file content which will contain the newer versions.
For each changes that the packager will do, a new build will trigger and then a new check_upgrade execution but honestly, I'd see well that as background job but i don't know if OBS have a similar mechanism.

**Comment by pandom79** [comment]
You can start to test from [here](https://github.com/pandom79/obs-service-check_upgrade).
There is also a _service file with some examples.
Obviously, This is a draft yet.

**Comment by pandom79** [comment]
Maybe [this](https://guides.rubyonrails.org/active_job_basics.html) could help you to do that ?
Each package could contains a "**_job**" file as "**_service**" in which we can define our parameters, therefore, we can use the services mechanism for the jobs too...
The "**_job**"  file could contain a xml similar to the services....In the future could be other jobs.
Am i dreaming ? :)
